graviti/openapi/policy.py

```diff
@@ -5,9 +5,101 @@
 
 """Interfaces about the dataset policy."""
 
-from typing import Any, Dict, Optional
+import base64
+import hashlib
+import hmac
+from datetime import datetime, timezone
+from typing import Any, ByteString, Dict, Optional, Tuple
 
-from graviti.openapi.requests import open_api_do
+import xmltodict
+
+from graviti.exception import ResponseError
+from graviti.openapi.requests import do, open_api_do
+from graviti.utility import config, convert_iso_to_datetime, locked
+
+_EXPIRED_IN_SECOND = 240
+_GMT_FORMAT = "%a, %d %b %Y %H:%M:%S GMT"
+_PERMISSIONS: Dict[Tuple[str, str, str, str], Dict[str, Any]] = {}
+_DEFAULT_PERMISSION = {"expireAt": datetime.fromtimestamp(0, timezone.utc)}
+
+
+@locked
+def _request_read_permission(
+    access_key: str,
+    url: str,
+    owner: str,
+    dataset: str,
+) -> None:
+    permission: Dict[str, Any] = get_object_policy(
+        access_key,
+        url,
+        owner,
+        dataset,
+        is_internal=config.is_internal,
+        expired=_EXPIRED_IN_SECOND,
+    )
+    permission["expireAt"] = convert_iso_to_datetime(permission["policy"].pop("expireAt"))
+    _PERMISSIONS[access_key, owner, dataset, "GET"] = permission
+
+
+def _get_read_permission(
+    access_key: str,
+    url: str,
+    owner: str,
+    dataset: str,
+    request_again: bool,
+) -> Dict[str, Any]:
+    key = (access_key, owner, dataset, "GET")
+    if (
+        request_again
+        or datetime.now(timezone.utc) >= _PERMISSIONS.get(key, _DEFAULT_PERMISSION)["expireAt"]
```

Comment:
will it be better to send request first, use the error message to check the expire issue, then ask the permission and rerequest